crates/history/src/query.rs

```diff
@@ -0,0 +1,114 @@
+//! # Encoding and decoding strategies for query strings.
+//!
+//! There are various strategies to map Rust types into HTTP query strings. The [`FromQuery`] and
+//! [`ToQuery`] encode the logic for how this encoding and decoding is performed. These traits
+//! are public as a form of dependency inversion, so that you can override the decoding and
+//! encoding strategy being used.
+//!
+//! These traits are used by the [`History`](crate::History) trait, which allows for modifying the
+//! history state, and the [`Location`](crate::Location) struct, which allows for extracting the
+//! current location (and this query).
+//!
+//! ## Default Strategy
+//!
+//! By default, any Rust type that implements [`Serialize`] or [`Deserialize`](serde::Deserialize)
+//! has an implementation of [`ToQuery`] or [`FromQuery`], respectively. This implementation uses
+//! the `serde_urlencoded` crate, which implements a standards-compliant `x-www-form-urlencoded`
+//! encoder and decoder. Some patterns are not supported by this crate, for example it is not
+//! possible to serialize arrays at the moment. If this is an issue for you, consider using the
+//! `serde_qs` crate.
+//!
+//! Example:
+//!
+//! ```rust,no_run
+//! use serde::{Serialize, Deserialize};
+//! use gloo_history::{MemoryHistory, History};
+//!
+//! #[derive(Serialize)]
+//! struct Query {
+//!     name: String,
+//! }
+//!
+//! let query = Query {
+//!     name: "user".into(),
+//! };
+//!
+//! let history = MemoryHistory::new();
+//! history.push_with_query("index.html", &query).unwrap();
+//! ```
+//!
+//! ## Custom Strategy
+//!
+//! If desired, the [`FromQuery`] and [`ToQuery`] traits can also be manually implemented on
+//! types to customize the encoding and decoding strategies. See the documentation for these traits
+//! for more detail on how this can be done.
+use crate::error::HistoryError;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{
+    borrow::Cow,
+    convert::{AsRef, Infallible},
+};
+
+/// Type that can be encoded into a query string.
+pub trait ToQuery {
+    /// Error that can be returned from the conversion.
+    type Error;
+
+    /// Method to encode the query into a string.
+    fn to_query(&self) -> Result<Cow<'_, str>, Self::Error>;
+}
+
+/// Type that can be decoded from a query string.
+pub trait FromQuery {
+    /// Target type after parsing.
+    type Target;
+    /// Error that can occur while parsing.
+    type Error;
```

Comment:
Is there a reason why `Error` doesn't have a [`std::error::Error`](https://doc.rust-lang.org/nightly/core/error/trait.Error.html) trait bound?

Reply:
I decided not to bound it here because we don't do anything with the error besides return it. If you were using this trait anywhere and you do care that the error type implements `std::error::Error`, you can always do:

```rust
fn do_something<T: FromQuery>() where T::Error: std::error::Error {}
```

If you use the APIs with anything that implements `Serialize` then the error type will be serialize to. So if you bound is `T: Serialize`, then it works as expected even without the `where` bound.

My rule of thumb is to keep traits as generic as possible — even if 99% of users would have an Error type that implements `std::error::Error`, unless I use that directly, I do not constrain it. One example is `anyhow::Error`, which does not implement `std::error::Error`  (for technical reasons) and yet is widely used. So adding a constraint of `type Error: std::error::Error` would make it impossible to use `anyhow::Error` as the error type (iirc).